### PR TITLE
feat: add upstream dns validation

### PR DESCRIPTION
**Pull request by arturshadnik**
## Issue
#84 

## Description
- added service for validating that at least a certain number of upstream DNS servers are configured
- added unit tests
- expanded integration tests to include new rule

### CODE REVIEW

File: internal/controller/maasvalidator_controller.go [hunk 1]
```diff
@@ -117,13 +118,23 @@ func (r *MaasValidatorReconciler) Reconcile(ctx context.Context, req ctrl.Reques
...
+	imageRulesService := osval.NewImageRulesService(r.Log, maasClient.BootResources)
+	upstreamDNSRulesService := dnsval.NewUpstreamDNSRulesService(r.Log, maasClient.MAASServer)
 
 	// MAAS Instance image rules
 	for _, rule := range validator.Spec.ImageRules {
-		vrr, err := maasRuleService.ReconcileMaasInstanceImageRule(rule)
+		vrr, err := imageRulesService.ReconcileMaasInstanceImageRule(rule)
 		if err != nil {
-			r.Log.V(0).Error(err, "failed to reconcile MAAS instance rule")
+			r.Log.V(0).Error(err, "failed to reconcile MAAS instance image rule")
```
Comment (TylerGillson) [inline internal/controller/maasvalidator_controller.go:128]: nit: "MAAS instance image rule" -> "MAAS image rule"

File: internal/controller/maasvalidator_controller.go [hunk 2]
```diff
@@ -117,13 +118,23 @@ func (r *MaasValidatorReconciler) Reconcile(ctx context.Context, req ctrl.Reques
...
+		}
+		resp.AddResult(vrr, err)
+	}
+
+	// MAAS Instance upstream DNS rules
+	for _, rule := range validator.Spec.UpstreamDNSRules {
+		vrr, err := upstreamDNSRulesService.ReconcileMaasInstanceUpstreamDNSRules(rule)
+		if err != nil {
+			r.Log.V(0).Error(err, "failed to reconcile MAAS instance upstream DNS rule")
```
Comment (TylerGillson) [inline internal/controller/maasvalidator_controller.go:137]: nit: "MAAS instance upstream DNS rule" -> "MAAS upstream DNS rule"

File: internal/validators/dns/upstream_dns_validator.go [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+	return details, errs
+}
+
+// buildValidationResult builds a default ValidationResult for a given validation type
+func buildValidationResult(rule v1alpha1.UpstreamDNSRule) *types.ValidationRuleResult {
+	state := vapi.ValidationSucceeded
+	latestCondition := vapi.DefaultValidationCondition()
+	latestCondition.Details = make([]string, 0)
+	latestCondition.Failures = make([]string, 0)
+	latestCondition.Message = fmt.Sprintf("All %s checks passed", constants.ValidationTypeImage)
```
Comment (TylerGillson) [inline internal/validators/dns/upstream_dns_validator.go:78]: should be constants.ValidationTypeUpstreamDNS

File: internal/validators/dns/upstream_dns_validator.go [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+	latestCondition.ValidationRule = fmt.Sprintf("%s-%s", vapiconstants.ValidationRulePrefix, util.Sanitize(rule.Name))
+	latestCondition.ValidationType = constants.ValidationTypeImage
+	return &types.ValidationRuleResult{Condition: &latestCondition, State: &state}
+}
+
+// updateResult updates a ValidationRuleResult with a list of errors and details
+func updateResult(vr *types.ValidationRuleResult, errs []error, errMsg string, details ...string) {
```
Comment (TylerGillson) [inline internal/validators/dns/upstream_dns_validator.go:85]: could this be a common function shared across the various validator modules?

**Review by TylerGillson — CHANGES_REQUESTED** [review]

File: internal/validators/dns/upstream_dns_validator.go (hunk 4 above)
Comment (arturshadnik) [reply internal/validators/dns/upstream_dns_validator.go:85]: yes, i'll be pulling this and `buildValidationResult()` out for reuse. Was planning it for a later PR but I will include it here instead.

**Pull request merged**